controllers/hcloudremediation_controller_test.go

```diff
@@ -0,0 +1,292 @@
+	})
+
+	Context("Basic test", func() {
+		It("creates the hcloudRemediation successfully", func() {
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() error {
+				return testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation)
+			}, timeout).Should(BeNil())
+		})
+
+		It("checks if hcloudRemediation objects has the HCloudTokenAvailableCondition condition", func() {
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				testEnv.GetLogger().Info("Status condition of the hcloud remediation", "condition", hcloudRemediation.Status.Conditions)
+				return isPresentAndTrue(hcloudRemediationkey, hcloudRemediation, infrav1.HCloudTokenAvailableCondition)
+			}, timeout).Should(BeTrue())
+		})
+
+		It("checks that no remediation is tried if HCloud server does not exist anymore", func() {
+			By("ensuring if hcloudMachine is provisioned")
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudMachineKey, hcloudMachine); err != nil {
+					return false
+				}
+
+				testEnv.GetLogger().Info("Status of the hcloudmachine", "status", hcloudMachine.Status)
+				return hcloudMachine.Status.Ready
+			}, timeout).Should(BeTrue())
+
+			By("deleting the server associated with the hcloudMachine")
+			providerID, err := hcloudutil.ServerIDFromProviderID(hcloudMachine.Spec.ProviderID)
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(hcloudClient.DeleteServer(ctx, &hcloud.Server{ID: providerID})).NotTo(HaveOccurred())
+
+			By("creating the hcloudRemediation")
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			By("checking if hcloudRemediation is in deleting phase and capiMachine has the MachineOwnerRemediatedCondition")
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				return hcloudRemediation.Status.Phase == infrav1.PhaseDeleting &&
+					isPresentAndFalseWithReason(capiMachineKey, capiMachine, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason)
+			}, timeout).Should(BeTrue())
+		})
+
+		It("checks that, under normal conditions, a reboot is carried out and retryCount and lastRemediated are set", func() {
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
+					return false
+				}
+
+				return hcloudRemediation.Status.LastRemediated != nil && hcloudRemediation.Status.RetryCount == 1
+			}, timeout).Should(BeTrue())
+		})
+
+		It("checks if PhaseWaiting is set when retryLimit is reached", func() {
+			Expect(testEnv.Create(ctx, hcloudRemediation)).To(Succeed())
+
+			Eventually(func() bool {
+				if err := testEnv.Get(ctx, hcloudRemediationkey, hcloudRemediation); err != nil {
```

Comment:
func () error

Reply:
why?? in the later line, I have checked if return `hcloudRemediation.Status.Phase == infrav1.PhaseWaiting` how does making it error will help?

Reply:
I mis-read the brackets below. My bad!